lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp

```diff
@@ -70,164 +32,325 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
+  auto origSmem = getSharedMemoryObjectFromStruct(
+      loc, llA, typeConverter->convertType(opTensorTy.getElementType()),
       rewriter);
-  Value strideBN = bSmem.strides[1];
-  Value strideBK = bSmem.strides[0];
-  Value strideB0 = isBRow ? strideBN : strideBK;
-  Value strideB1 = isBRow ? strideBK : strideBN;
-  int bNumPtr = 8;
-  int K = bShapePerCTA[0];
-  int N = bShapePerCTA[1];
-
-  auto shapePerCTATile = getShapePerCTATile(dLayout);
-  auto sizePerThread = getSizePerThread(dLayout);
-
-  Value _0 = i32_val(0);
-
-  Value nContig = i32_val(sizePerThread[order[0]]);
-
-  // threadId in blocked layout
-  auto threadIds = getThreadIds(thread, shapePerCTATile, sizePerThread, order,
-                                rewriter, loc);
-  Value threadIdN = threadIds[1];
-
-  Value offB0 = isBRow ? mul(threadIdN, nContig) : _0;
-  Value offB1 = isBRow ? _0 : mul(threadIdN, nContig);
-  SmallVector<Value> bOff(bNumPtr);
-  for (int i = 0; i < bNumPtr; ++i) {
-    bOff[i] = add(mul(offB0, strideB0), mul(offB1, strideB1));
+  auto smem = getExpandedSharedMemoryObject(rewriter, loc, origSmem,
+                                            opTensorTy.getShape());
+  auto strides = smem.strides;
+  int B = opTensorShape[dim.batch];
+  int K = opTensorShape[dim.k];
+  int NonK = opTensorShape[dim.nonK];
+
+  auto shapePerCTATile =
+      expandMatrixShapeWithBatch(ArrayRef(getShapePerCTATile(dLayout)));
+  shapePerCTATile[dim.k] = K;
+  auto sizePerThread =
+      expandMatrixShapeWithBatch(ArrayRef(getSizePerThread(dLayout)));
+  sizePerThread[dim.k] = K;
+  auto threadsPerWarp =
+      expandMatrixShapeWithBatch(ArrayRef(dLayout.getThreadsPerWarp()));
+  auto warpsPerCTA =
+      expandMatrixShapeWithBatch(ArrayRef(dLayout.getWarpsPerCTA()));
+
+  auto warpSize = i32_val(triton::gpu::getWarpSize(dLayout));
+  auto laneId = urem(thread, warpSize);
+  auto warpId = udiv(thread, warpSize);
+  auto laneIds =
+      mlir::LLVM::delinearize(rewriter, loc, laneId, threadsPerWarp, opOrder);
+  auto warpIds =
+      mlir::LLVM::delinearize(rewriter, loc, warpId, warpsPerCTA, opOrder);
+  auto sizePerWarpB = sizePerThread[dim.batch] * threadsPerWarp[dim.batch];
+  auto sizePerWarpNonK = sizePerThread[dim.nonK] * threadsPerWarp[dim.nonK];
+
+  Value bTileOffset =
+      mul(laneIds[dim.batch], i32_val(sizePerThread[dim.batch]));
+  bTileOffset =
+      add(bTileOffset, mul(warpIds[dim.batch], i32_val(sizePerWarpB)));
+  Value nonKTileOffset =
+      mul(laneIds[dim.nonK], i32_val(sizePerThread[dim.nonK]));
+  nonKTileOffset =
+      add(nonKTileOffset, mul(warpIds[dim.nonK], i32_val(sizePerWarpNonK)));
+
+  auto elemTy = typeConverter->convertType(opTensorTy.getElementType());
+  Type ptrTy = smem.base.getType();
+
+  auto sharedOrder = expandMatrixOrderWithBatch(sharedLayout.getOrder());
+  // compute contiguity of fastest dimension in shared layout.
+  unsigned vectorSize = sizePerThread[sharedOrder[0]];
```

Comment:
Let's say for opIdx=0, and K = 16. This will lead to vectorSize = 16, which is larger than the max vectorSize of a single ds_read instruction. 
In the current implementation, K=16 will go to the mfma path. But in the future, we want to to try fma for small-M but larger K. So it's better to restrict vectorSize <= 128B/elemSize

Reply:
This is a common code base. Let's think about this in next PRs.
Maybe add a new functions "getMaxSharedLoadWidth" function in target info?